Allow user cancellation of DIFM migrations

markbiek opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to DOTCON-26-linear-issue

## Proposed Changes

This adds a button to the site overview in the case where the user has a DIFM migration started.

Clicking the button displays a dialog which lets the user send a request to cancel the migration. This calls the
/automated-migration/cancel-migration endpoint.

We display a success or error notice depending on the result from the endpoint.

If the request is successful, the user will see the normal site overview on refresh.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Use the calypso.live url or run this locally. NOTE: If you run it locally, you'll need to share credentials instead of authorizing the application password once you reach that step.
* Go to `/start` and continue to the Goals step.
* Select the 'import or migrate' intent.
* Continue with the migration flow and select a DIFM migration.
* When you reach the site overview, there should be a new 'Request cancellation' button. 
![image](https://github.com/user-attachments/assets/c5a8a8c6-1010-41ab-87ce-47e4365b12ce)
  * If you don't see the button, try enabling the `migration-flow/cancel-difm` flag.
* Clicking the button should display the following dialog.
<img width="542" alt="CleanShot 2025-05-07 at 16 17 51@2x" src="https://github.com/user-attachments/assets/6b5f7df8-1847-4e8f-a28d-6d9a5726e9f1" />

* Clicking 'Send Request' should display a success notice.
 
![image](https://github.com/user-attachments/assets/34319f2f-bee9-497e-b46f-f2ac2f156c60)

* Once the migration has been cancelled, find the Zendesk ticket for the site. The ticket should have a 'The user has requested to cancel the migration.' internal reply.
<img width="433" alt="CleanShot 2025-05-06 at 14 01 28@2x" src="https://github.com/user-attachments/assets/2143bec0-fb15-4a14-bc70-909b7e20753f" />

In the unlikely event of an error, it should display an error notice:
<img width="959" alt="CleanShot 2025-05-08 at 09 52 36@2x" src="https://github.com/user-attachments/assets/7de97867-d542-4031-ae55-c48d34930127" />


You can simulate an error by sandboxing the public api and modifying the cancellation endpoint (site-automated-migration.php#L442) as follows.

You may need to reset the site by removing the `migration-cancelled-difm` sticker and adding the `migration-started-difm` sticker.

You'll want to add the following to your `0-sandbox.php` to direct any Slack error messages to a personal channel.
```
add_filter(
	'wpcom_migration_notification_slack_channel',
	function( $channel ) {
		return 'CHANNEL_ID';
	}
);
```
You can get the channel id (I like to use the Slackbot channel) by clicking on the channel and clicking the `...` in the upper right, then clicking "View full profile".
![image](https://github.com/user-attachments/assets/df80aace-738e-4b57-a1c1-e43b38c21177)


Then click the `...` on the profile and click "Copy member id".
![image](https://github.com/user-attachments/assets/f809e95f-aa6c-4463-b740-d1dfeec8c647)


Then change this code
```
if ( ! $current_site ) {
			self::automated_migration_notification_slack(
				(int) $request->get_param( 'wpcom_site' ) ?? -1,
				wp_get_current_user(),
				$cancellation_error_message,
				array(
					'error' => 'Unable to get blog details',
				)
			);
			return new WP_Error( 'site_not_found', 'Site not found.', array( 'status' => WP_Http::NOT_FOUND ) );
		}
```

to

```
if ( true || ! $current_site ) {
	return new WP_Error( 'site_not_found', 'Site not found.', array( 'status' => WP_Http::NOT_FOUND ) );
}
```


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

markbiek committed:
Allow user cancellation of DIFM migrations

This adds a button to the site overview in the case where the user has a
DIFM migration started.

Clicking the button displays a dialog which lets the user send a request
to cancel the migration. This calls the
/automated-migration/cancel-migration endpoint.

We display a success or error notice depending on the result from the
endpoint.

If the request is successful, the user will see the normal site overview
on refresh.

fditrapani commented:
Thanks for working on this Mark. I have few minor UI related things:

1. Can we use the [core component for the modal](https://wordpress.github.io/gutenberg/?path=/docs/components-modal--docs). We've been making a push to standardize it across the board.
2. Can we use a [Global notice](https://wpcalypso.wordpress.com/devdocs/design/global-notices) instead of an inline one for the success and error messages. It feels weird to display them inline like that. We might want to tweak the copy so they don't appear so long.

markbiek commented:
> Thanks for working on this Mark. I have few minor UI related things:
> 
>     1. Can we use the [core component for the modal](https://wordpress.github.io/gutenberg/?path=/docs/components-modal--docs). We've been making a push to standardize it across the board.
> 
>     2. Can we use a [Global notice](https://wpcalypso.wordpress.com/devdocs/design/global-notices) instead of an inline one for the success and error messages. It feels weird to display them inline like that. We might want to tweak the copy so they don't appear so long.

Thanks for those notes! I wasn't aware of the core modal but it should be pretty straightforward to switch. I'll take care of that and the notifications.

And any copy suggestions are very welcome!

fditrapani commented:
Awesome! Thanks. Can you share screenshots with the new notices. I'll see how we might tweak those from there!

markbiek commented:
@fditrapani I pushed a change to use `Modal` instead of `Dialog`. I've updated the screenshot in the description so let me know if you have any tweaks.

I'll let you know when I've updated the notices.

markbiek commented:
@fditrapani I've updated it to use the `GlobalNotice` instead. Screenshots are updated in the PR description, let me know if you have any copy tweaks.

fditrapani commented:
Awesome! We're getting there!

- Can we use the `medium` size on the modal for a shorter line length to help with readability.
- Can we use this for the error notice "We couldn’t cancel your migration. Our support team will reach out to help."

markbiek commented:
>     * Can we use the `medium` size on the modal for a shorter line length to help with readability.
> 
>     * Can we use this for the error notice "We couldn’t cancel your migration. Our support team will reach out to help."

@fditrapani I've made both of those changes (screenshots are updated in the PR description).

fditrapani reviewed (approved):
Amazing. Thank you! Looks great on my end.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17455147

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @markbiek for including a screenshot in the description! This is really helpful for our translators.
